### PR TITLE
Optimize fptrunc(x)>=C1 -->  x>=C2

### LINKED ISSUE #85265: missed optimization, fptrunc (x) >= C1 => x >= C2
https://alive2.llvm.org/ce/z/n53up6
https://godbolt.org/z/q6vrbxTo3
```llvm
define i1 @src(double %0) {
if.end155:
%1 = fptrunc double %0 to float
%2 = fcmp oge float %1, 1.000000e+02
ret i1 %2
}
define i1 @tgt(double %0) {
if.end155:
%oge = fcmp oge double %0, 0x4058FFFFF0000000
ret i1 %oge
}
```

Other variants may follow the same patterns.

### PR DESCRIPTION
**Pull request by kissholic**
Fix https://github.com/llvm/llvm-project/issues/85265#issue-2186848949

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Please read the guideline https://llvm.org/docs/InstCombineContributorGuide.html.

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+
+;CHECK-LABEL: @src(
+;CHECK: %result = fcmp oge double %0, 1.000000e+02
+;CHECK-NEXT: ret i1 %result
+define i1 @src(double %0) {
+    %trunc = fptrunc double %0 to float
+    %result = fcmp oge float %trunc, 1.000000e+02
+    ret i1 %result
+}
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:11]: Missing newline.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -7882,6 +7885,30 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+
+  // Check that predicates are valid.
+  if ((Pred != FCmpInst::FCMP_OGT) && (Pred != FCmpInst::FCMP_OLT) &&
+      (Pred != FCmpInst::FCMP_OGE) && (Pred != FCmpInst::FCMP_OLE))
+    return nullptr;
+
+  auto *LType = LHSI->getOperand(0)->getType();
+  auto *RType = RHSC->getType();
+
+  if (!(LType->isFloatingPointTy() && RType->isFloatingPointTy() &&
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7902]: You shouldn't need to check isFloatingPointTy, that's implied by the operations in use already. Also not sure what a >= means when comparing type IDs but you probably don't need that either

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -7882,6 +7885,30 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+        LType->getTypeID() >= RType->getTypeID()))
+    return nullptr;
+
+  auto *ROperand = llvm::ConstantFP::get(
+      LType, dyn_cast<ConstantFP>(RHSC)->getValue().convertToDouble());
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7907]: Don't use convertToDouble, keep this entirely in APFloat

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll (hunk 1 above)
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:11]: Test more combinations of types, including vectors. Also test flag preservation behavior

**Comment by aengelke** [comment]
This appears to be incorrect w.r.t. round-to-nearest rounding of fptrunc. [alive2](https://alive2.llvm.org/ce/z/w-496p) The constant needs adjustment.

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll (hunk 1 above)
Comment (jcranmer-intel) [reply llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:11]: A test of `fp128`, `x86_fp80`, or `ppc_fp128` in particular would be helpful.

Comment (kissholic) [reply llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:11]: > Test more combinations of types, including vectors. Also test flag preservation behavior
@arsenm Sorry, could you give me a hint what 'flag preservation behavior' means in IR?

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:11]: I mean the flags on the compare should be preserved, and you don't have any tests using fast math flags. e.g. https://alive2.llvm.org/ce/z/uQr4-J

Comment (kissholic) [reply llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:11]: > A test of `fp128`, `x86_fp80`, or `ppc_fp128` in particular would be helpful.

Sorry for late. 

I tried to test these types, but an error is generated whose message is "floating point constant does not have type 'fp128'". 

The error is emitted in the parse stage, and more modifications might be required to be conducted.

Should i ignore this problem, or if there are better solutions?

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:11]: The IR syntax for floating point constants is unnecessarily painful. You need to use a different prefix and have the correct hex length depending on the format. For this I usually just write fpext from a reasonable FP type to the long one, and run it through instsimplify to see how it should be printed

**Comment by kissholic** [comment]
> This appears to be incorrect w.r.t. round-to-nearest rounding of fptrunc. [alive2](https://alive2.llvm.org/ce/z/w-496p) The constant needs adjustment.

It seems that the double type the fp constant converted to can express the same value with float type without lossing accuracy. The rmNearestTiesToEven has already been applied, and no difference appeared.🫣

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 4]
```diff
@@ -7888,25 +7894,48 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
-  if (!(LType->isFloatingPointTy() && RType->isFloatingPointTy() &&
-        LType->getTypeID() >= RType->getTypeID()))
-    return nullptr;
+    return new FCmpInst(Pred, LHSI->getOperand(0),
+                        ConstantFP::get(LType, RValue), "", &I);
+  }
+
+  if (RHSC->getType()->isVectorTy()) {
+    Type *LVecType = LHSI->getOperand(0)->getType();
+    Type *LEleType = dyn_cast<VectorType>(LVecType)->getElementType();
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7917]: unchecked dyn_cast. Use the dyn_cast in the if expression instead of using isVectorTy

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 5]
```diff
@@ -7888,25 +7894,48 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+
+    FixedVectorType *VecType = dyn_cast<FixedVectorType>(RHSC->getType());
+    uint64_t EleNum = VecType->getNumElements();
 
-  auto *ROperand = llvm::ConstantFP::get(
-      LType, dyn_cast<ConstantFP>(RHSC)->getValue().convertToDouble());
+    std::vector<Constant *> EleVec(EleNum);
+    for (uint64_t Idx = 0; Idx < EleNum; ++Idx) {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7923]: This is trying too hard to handle the non-splat case. Just use m_APFloat which handles splat vectors and scalars at the same time

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll [hunk 6]
```diff
@@ -1,11 +1,125 @@
...
+; CHECK-LABEL: define <4 x i1> @fcmp_vec_trunc_with_flag(
+; CHECK-SAME: <4 x double> [[TMP0:%.*]]) {
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp fast olt <4 x double> [[TMP0]], <double 1.000000e+00, double 2.000000e+00, double 3.000000e+00, double 4.000000e+00>
+; CHECK-NEXT:    ret <4 x i1> [[CMP]]
+;
+  %vec = fptrunc <4 x double> %0 to <4 x float>
+  %cmp = fcmp fast olt <4 x float> %vec, <float 1.0, float 2.0, float 3.0, float 4.0>
+  ret <4 x i1> %cmp
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:92]: Add a scalable vector test

**Comment by aengelke** [comment]
> It seems that the double type the fp constant converted to can express the same value with float type without lossing accuracy. The rmNearestTiesToEven has already been applied, and no difference appeared.🫣

It's the opposite direction that is problematic. Consider input `99.99999999`. After the fptrunc, it will be value `100.0f`, which is >= 100.0f. But `99.99999999` is not >= `100.0`. You need to find the smallest(/largest) value of the larger floating-point type which, after truncation, is satisfies the condition.

The [contributor guide](https://llvm.org/docs/InstCombineContributorGuide.html) also says that you should provide alive2 proofs that your transformation is correct. I provided a proof above that the transformation as implemented/tested now is not correct.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 5 above)
Comment (kissholic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7923]: So i should split splat vector case from the normal cases, and combine it with scalars, if i understand correctly?😖

Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7923]: You shouldn't have to really worry about the vector case. If you use m_APFloat, it should just work. It will handle scalars and splat vectors

Comment (kissholic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7923]: > You shouldn't have to really worry about the vector case. If you use m_APFloat, it should just work. It will handle scalars and splat vectors

Sorry, i still didn't get the point... It seems that m_APFloat can't cover the non-splat vector cases.

I also read the icmp-trunc optimization code and ran an int vector test case, but it seems not work in the int vector case.

I came up with an idea that FPExtInst may be applied to the constant, and left the optimization to constant extension part 😋 (joking).

Could you give some more hints? Thank you <3

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 7]
```diff
@@ -7882,6 +7892,79 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
   return new FCmpInst(Pred, LHSI->getOperand(1), RHSC, "", &I);
 }
 
+// Fold trunc(x) < constant --> x < constant if possible.
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7895]: fptrunc, not trunc

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 8]
```diff
@@ -7882,6 +7892,79 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+static Instruction *foldFCmpFpTrunc(FCmpInst &I, Instruction *LHSI,
+                                    Constant *RHSC) {
+  FCmpInst::Predicate Pred = I.getPredicate();
+  bool RoundDown = false;
+
+  if ((Pred == FCmpInst::FCMP_OGE) || (Pred == FCmpInst::FCMP_UGE) ||
+      (Pred == FCmpInst::FCMP_OLT) || (Pred == FCmpInst::FCMP_ULT))
+    RoundDown = true;
+  else if ((Pred == FCmpInst::FCMP_OGT) || (Pred == FCmpInst::FCMP_UGT) ||
+           (Pred == FCmpInst::FCMP_OLE) || (Pred == FCmpInst::FCMP_ULE))
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7905]: Don't need parentheses around all these == expressions

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+; CHECK-LABEL: define i1 @fcmp_trunc_ppc_fp128(
+; CHECK-SAME: ppc_fp128 [[TMP0:%.*]]) {
+; CHECK-NEXT:    [[RESULT:%.*]] = fcmp fast oge ppc_fp128 [[TMP0]], 0xM4058FFFFF0000000BD00000000000000
+; CHECK-NEXT:    ret i1 [[RESULT]]
+;
+  %trunc = fptrunc ppc_fp128 %0 to float
+  %result = fcmp fast oge float %trunc, 1.000000e+02
+  ret i1 %result
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:169]: The set of tested constants seems too simple for the complexity of the loop testing constant validity. Should have negative tests for off by one bit in each direction. Also test with the edge case constants (inf, nan) and some denormal values?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 10]
```diff
@@ -7882,6 +7892,79 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+
+  APFloat RoundValue{LEleType->getFltSemantics()};
+  {
+    APFloat Two{LEleType->getFltSemantics(), 2};
+    APFloat LowBound = RoundDown ? ExtNextRValue : ExtRValue;
+    APFloat UpBound = RoundDown ? ExtRValue : ExtNextRValue;
+
+    while (true) {
+      APFloat DupUpBound = UpBound;
+      DupUpBound.next(true);
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7939]: Name UpBoundNext, or similar?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 11]
```diff
@@ -7882,6 +7892,79 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+  ExtNextRValue.convert(LEleType->getFltSemantics(),
+                        APFloat::rmNearestTiesToEven, &lossInfo);
+
+  APFloat RoundValue{LEleType->getFltSemantics()};
+  {
+    APFloat Two{LEleType->getFltSemantics(), 2};
+    APFloat LowBound = RoundDown ? ExtNextRValue : ExtRValue;
+    APFloat UpBound = RoundDown ? ExtRValue : ExtNextRValue;
+
+    while (true) {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7937]: This loop is the hard to review part and needs some comments explaining what constants are legal

**Review by arsenm — COMMENTED** [review]
Can you add the alive2 link proof to the description

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 11 above)
Comment (kissholic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7937]: Sorry for late, being occupied by work.

The alive proof will be posted soon. 

The edge cases (inf, nan) has been folded before entering this optimization (e.g. fcmp oge x inf --> fcmp oeq x inf). Maybe filtering these cases is a good idea?

Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7937]: It's not always safe to rely on those folding before hand

**Comment by kissholic** [comment]
alive2 proof:
https://alive2.llvm.org/ce/z/rphqKP
https://alive2.llvm.org/ce/z/UhU5nG
https://alive2.llvm.org/ce/z/v3zu93
https://alive2.llvm.org/ce/z/fQxsaA
https://alive2.llvm.org/ce/z/UYnaSb
https://alive2.llvm.org/ce/z/5VRZGg

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+; CHECK-LABEL: define i1 @fcmp_trunc_mn(
+; CHECK-SAME: double [[TMP0:%.*]]) {
+; CHECK-NEXT:    [[RESULT:%.*]] = fcmp olt double [[TMP0]], 0xC7EFFFFFEFFFFFFF
+; CHECK-NEXT:    ret i1 [[RESULT]]
+;
+  %trunc = fptrunc double %0 to float
+  %result = fcmp olt float %trunc, -3.4028234663852885981170418348451692544e38
+  ret i1 %result
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fold-fcmp-trunc.ll:302]: Test the literal nan and inf cases

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 13]
```diff
@@ -7882,6 +7892,93 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+  FCmpInst::Predicate Pred = I.getPredicate();
+  bool RoundDown = false;
+
+  if (Pred == FCmpInst::FCMP_OGE || Pred == FCmpInst::FCMP_UGE ||
+      Pred == FCmpInst::FCMP_OLT || Pred == FCmpInst::FCMP_ULT)
+    RoundDown = true;
+  else if (Pred == FCmpInst::FCMP_OGT || Pred == FCmpInst::FCMP_UGT ||
+           Pred == FCmpInst::FCMP_OLE || Pred == FCmpInst::FCMP_ULE)
+    RoundDown = false;
+  else
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7907]: uge, ule cases not tested. Plus negative test for others

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 14]
```diff
@@ -7882,6 +7892,93 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+  // allowed here.
+  APFloat RoundValue{LEleType->getFltSemantics()};
+  {
+    APFloat Two{LEleType->getFltSemantics(), 2};
+
+    // The (negative) maximum of RValue will become infinity when rounded up (down).
+    // Set the limit of ExtNextRValue.
+    if (NextRValue.isInfinity()) {
+      ExtNextRValue = ExtRValue * Two;
+    }
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7946]: No braces. Can also defer construction of the 2 constant (or avoid it by using scalbn instead)

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 15]
```diff
@@ -7882,6 +7892,93 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+  // Round RValue to suitable value
+  APFloat ExtRValue = *RValue;
+  APFloat ExtNextRValue = NextRValue;
+  bool lossInfo;
+  ExtRValue.convert(LEleType->getFltSemantics(), APFloat::rmNearestTiesToEven,
+                    &lossInfo);
+  ExtNextRValue.convert(LEleType->getFltSemantics(),
+                        APFloat::rmNearestTiesToEven, &lossInfo);
+
+  // Binary search to find the maximal (or minimal) value after RValue promotion.
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7935]: Can you write this with std::lower_bound/std::upper_bound?

Comment (kissholic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7935]: std::lower_bound/std::upper_bound (or similar algorithm like std::binary_search) seems only accept [ForwardIt type](https://en.cppreference.com/w/cpp/algorithm/lower_bound) and there is likely no suitable substitution algorithm in LLVM too, which may requires constructing a new complex wrapper struct of APFloat, such as implementing name required iteration methods, calculating the mean of two APFloat (without constructing a new APFloat divisor), defining a comparation function and so on.

Considering the internal complexity of APFloat wrapper, it is simpler to keep the original one (i think)?

**Comment by kissholic** [comment]
More test cases are added, and the refactor partial logics of 'foldFCmpFpTrunc'

https://alive2.llvm.org/ce/z/nodhVp
https://alive2.llvm.org/ce/z/PtGyK-
https://alive2.llvm.org/ce/z/QN4uVV
https://alive2.llvm.org/ce/z/GTD7r2
https://alive2.llvm.org/ce/z/JUPPTp
https://alive2.llvm.org/ce/z/yubkuX

**Comment by arsenm** [comment]
Why was this closed?

**Comment by kissholic** [comment]
> Why was this closed?

Sorry, the old commits seem to be blocked due to a merge operation. I tried to discard those commits and push it again.

**Comment by kissholic** [comment]
ping

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 16]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
   return new FCmpInst(Pred, LHSI->getOperand(1), RHSC, "", &I);
 }
 
+
+// Fold fptrunc(x) < constant --> x < constant if possible.
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]:
```suggestion
/// Fold fptrunc(x) < constant --> x < constant if possible.
```

This also covers other compare types, can you express that in the comment

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 17]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+static Instruction *foldFCmpFpTrunc(FCmpInst &I, Instruction *LHSI,
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]:
```suggestion
static Instruction *foldFCmpFpTrunc(FCmpInst &I, const Instruction &FPTrunc,
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 18]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+
+  const APFloat *RValue;
+  if (!match(RHSC, m_APFloat(RValue)))
+    return nullptr;
+
+  // RHSC should not be nan or infinity.
+  if (RValue->isNaN() || RValue->isInfinity())
+    return nullptr;
+
+  Type *LType = LHSI->getOperand(0)->getType();
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Seems more like a "Src" than a "Left" type

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 19]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+  Type *RType = RHSC->getType();
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Seems more like a "Dest" than a "Right" type

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 20]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+  if (NextRValue.isInfinity())
+    ExtNextRValue = scalbn(ExtRValue, 1, APFloat::rmNearestTiesToEven);
+
+  // Binary search to find the maximal (or minimal) value after 'RValue'
+  // promotion. 'RValue' should obey normal comparison rules, which means nan or
+  // inf is not allowed here.
+  APFloat RoundValue{LEleType->getFltSemantics()};
+
+  APFloat LowBound = RoundDown ? ExtNextRValue : ExtRValue;
+  APFloat UpBound = RoundDown ? ExtRValue : ExtNextRValue;
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]:
```suggestion
  const APFloat &LowBound = RoundDown ? ExtNextRValue : ExtRValue;
  const APFloat &UpBound = RoundDown ? ExtRValue : ExtNextRValue;
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 21]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+    APFloat UpBoundNext = UpBound;
+    UpBoundNext.next(true);
+    return LowBound == UpBoundNext;
+  };
+
+  auto EqualRValueAfterTrunc = [&](const APFloat &ExtValue) {
+    APFloat TruncValue = ExtValue;
+    TruncValue.convert(REleType->getFltSemantics(),
+                       APFloat::rmNearestTiesToEven, &lossInfo);
+    return TruncValue == *RValue;
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Do these compares amount to the same as directly using losesInfo?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 22]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+                        APFloat::rmNearestTiesToEven, &lossInfo);
+
+  // The (negative) maximum of 'RValue' may become infinity when rounded up
+  // (down). Set the limit of 'ExtNextRValue'.
+  if (NextRValue.isInfinity())
+    ExtNextRValue = scalbn(ExtRValue, 1, APFloat::rmNearestTiesToEven);
+
+  // Binary search to find the maximal (or minimal) value after 'RValue'
+  // promotion. 'RValue' should obey normal comparison rules, which means nan or
+  // inf is not allowed here.
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Move this comment down to the loop

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 23]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+  APFloat RoundValue{LEleType->getFltSemantics()};
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Move this down to where it's used before the loop

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 24]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+      break;
+    }
+
+    APFloat Mid = scalbn(LowBound + UpBound, -1, APFloat::rmNearestTiesToEven);
+    bool EqualRValue = EqualRValueAfterTrunc(Mid);
+
+    // 'EqualRValue' indicates whether Mid is qualified to be the final round
+    // value. if 'EqualRValue' == true, 'Mid' might be the final round value
+    //     if 'RoundDown' == true, 'UpBound' can't be the final round value
+    //     if 'RoudnDown' == false, 'DownBound' can't be the final round value
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]:
```suggestion
    //     if 'RoundDown' == false, 'DownBound' can't be the final round value
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 25]
```diff
@@ -8031,6 +8031,101 @@ static Instruction *foldFCmpReciprocalAndZero(FCmpInst &I, Instruction *LHSI,
...
+    return LowBound == UpBoundNext;
+  };
+
+  auto EqualRValueAfterTrunc = [&](const APFloat &ExtValue) {
+    APFloat TruncValue = ExtValue;
+    TruncValue.convert(REleType->getFltSemantics(),
+                       APFloat::rmNearestTiesToEven, &lossInfo);
+    return TruncValue == *RValue;
+  };
+
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8101]: I feel like this should be a directly computable constant without a binary search

**Review by arsenm — COMMENTED** [review]
The hard part of reviewing this is the binary search. I feel like this should be a directly computable value? I guess it's fine though, mostly nit comments

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 25 above)
Comment (kissholic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8101]: Yes, binary search can be replaced by average value for most cases, and I got confused before. However, it seems a little tricky to tackle the comparison with the largest value. I tried several ways to work around, but all are unsatisfactory. Might be late for the final solution.

Comment (kissholic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8101]: Although the absence of `next(max_float)` which is infinity actually, we can image its existence in wider types, and it should equal to `max_float + (max_float - prev(max_float))` in most cases (if not all). A bound check can ensure its correctness.

This method balance the efficiency and correctness, and keep consistence with logics for other floats despite the fact that it might not be standard for all float types.

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc-tmp.ll [hunk 26]
```diff
@@ -0,0 +1,596 @@
...
+; CHECK-NEXT:    [[RESULT:%.*]] = fcmp ogt double [[TMP0]], 0x3690000000000000
+; CHECK-NEXT:    ret i1 [[RESULT]]
+;
+  %trunc = fptrunc double %0 to float
+  %result = fcmp ogt float %trunc, 0x8000000000000000
+  ret i1 %result
+}
+
+
+; max representable
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fold-fcmp-trunc-tmp.ll]: Also test negative of max representable?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 25 above)
Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8101]: Are you planning on switching the patch to do this? Needs comments either way

Comment (kissholic) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8101]: I'm trying to, but it seems that the rule can't be applied to `ppc_fp128` when converted from `max_float` due to unknown precision loss. I want more investigation before decision

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc-tmp.ll (hunk 26 above)
Comment (kissholic) [reply llvm/test/Transforms/InstCombine/fold-fcmp-trunc-tmp.ll]: This case is covered by `fcmp_trunc_mn`. Maybe this name is misleading and should be changed.

**Comment by kissholic** [comment]
https://alive2.llvm.org/ce/z/bTf3ZW
Converting from `max float` to` ppc_fp128 ` will be `0xM47EFFFFFF0000000C150000000000000` which should pass the last finite check in the commit but fails. This error seems to be coming from the precision loss of addition of two IEEE 64-bit doubles inside of `ppc_fp128`.  This case will be bypassed temporarily before the issue fixed.

File: llvm/test/Transforms/InstCombine/fold-fcmp-trunc-tmp.ll (hunk 26 above)
Comment (kissholic) [reply llvm/test/Transforms/InstCombine/fold-fcmp-trunc-tmp.ll]: New algorithm is adopted~

**Review by arsenm — APPROVED** [review]
